### PR TITLE
Eclipse Integration E4

### PR DESCRIPTION
**Pull request by N1k145**
This is a first shot for an E4 Implementation https://github.com/eclipse-glsp/glsp/issues/1059

This is currently not fully functional ~, the context menu is currently not working~.
And the Content of the E4 Part is a Hard-coded file

What I did is, I moved alot of the Logic from the GLSPDiagramEditor class into the GLSPDiagramComposite and added an E4 Part called GLSPDiagramPart.

I would like some feedback on what you think about this approach.

### CODE REVIEW

**Comment by N1k145** [comment]
@planger Can you take a look, when you have time, if this goes in a direction that you can agree on.

**Comment by planger** [comment]
@N1k145 Thank you very much for this proposal! On a first glance this looks pretty good. We'll look at it in more detail in the next few days and get back to you. Thanks again!

**Comment by planger** [comment]
Hi @N1k145 again, I had another look at the change and I think this goes into a good direction. My only concern is that this change causes API breaks and it is rather short before the next planned major release (end of August). So I assume it'll not be possible to finish this work in the next week to enable making this part of the next major release.

Do you think there is a chance to make the e4 compatibility just adding API but not breaking any? This way, we could put it into the next minor release as it doesn't break but just add API.

**Comment by N1k145** [comment]
@planger this depends a bit on how much code duplication you would like to have.
The main issue is that currently a lot of the implementation depends hard on the GLSPDiagramEditor class.
Also, all the Setup Logic is hard in that class and depends partly on the Editor Part.

We could define an Interface that is implemented by the GLSPDiagramEditor and the GLSPDiagramComposite that provides the functionality the implementations around the Editor depends on. 

But moving the setup logic out of the Editor into the Composite is necessary, what we could do is duplicate the code, have it once in the Editor (without breaks) and the newer implementation in the Composite + E4 Part. That way we would add the two apis (Part and Composite) while leaving the Editor as is for now. And move that to the composite in the next major where breaking changes can be made.

**Comment by N1k145** [comment]
@planger I discussed this internally and I can spend more time on this in the next week. 
Are there any concrete changes you would like to have, or just cleaning it up and making it round?

**Comment by planger** [comment]
Ok cool, thanks for the feedback!
No I don't have any concrete changes, looks good so far. So cleaning up, removing all hard-coded parts and making the examples work as before (including the [project template](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse)) with all features (navigation, problem markers, shortcuts, etc.) is all that'd be required before we could merge it.

If we manage to get this done next week, we are happy to review it the week after. If it doesn't fit, I guess we could merge it after the release as an addition, keeping the editor class, but moving out code as much as possible into a class that is shared with the part implementation.

Thank you for your work! :+1:

File: server/example/org.eclipse.glsp.ide.workflow.editor/.checkstyle [hunk 1]
```diff
@@ -11,4 +11,14 @@
...
+    <property name="checkstyle.header.file" value="D:\dev\eclipse\eclipse-glsp-eclipse-integration\ws\.metadata\.plugins\org.eclipse.core.resources\.projects\org.eclipse.glsp.ide.workflow.editor\com.basistech.m2e.code.quality.checkstyleConfigurator\checkstyle-header-default.txt"/>
+    <property name="checkstyle.cache.file" value="${project_loc}/target/checkstyle-cachefile"/>
+  </local-check-config>
+  <fileset name="java-sources-default" enabled="true" check-config-name="maven-checkstyle-plugin default" local="true">
+    <file-match-pattern match-pattern="^src/.*\.java" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src.*.*src-gen/.*\.java" include-pattern="false"/>
+    <file-match-pattern match-pattern="^src/main/resources.*\.properties" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src/test/resources.*\.properties" include-pattern="true"/>
+  </fileset>
```
Comment (planger) [inline server/example/org.eclipse.glsp.ide.workflow.editor/.checkstyle:23]: I think this change is not really required. Also it contains absolute paths to your workspace.

File: server/example/org.eclipse.glsp.ide.workflow.editor/.classpath [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
-	<classpathentry kind="src" path="src/"/>
-	<classpathentry kind="output" path="target/classes"/>
-</classpath>
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/"/>
+	<classpathentry kind="output" path="target/classes"/>
+</classpath>
```
Comment (planger) [inline server/example/org.eclipse.glsp.ide.workflow.editor/.classpath:7]: This change seems to be unnecessary.

File: server/plugins/org.eclipse.glsp.ide.editor/.checkstyle [hunk 3]
```diff
@@ -11,4 +11,14 @@
...
+    <property name="checkstyle.header.file" value="D:\dev\eclipse\eclipse-glsp-eclipse-integration\ws\.metadata\.plugins\org.eclipse.core.resources\.projects\org.eclipse.glsp.ide.editor\com.basistech.m2e.code.quality.checkstyleConfigurator\checkstyle-header-default.txt"/>
+    <property name="checkstyle.cache.file" value="${project_loc}/target/checkstyle-cachefile"/>
+  </local-check-config>
+  <fileset name="java-sources-default" enabled="true" check-config-name="maven-checkstyle-plugin default" local="true">
+    <file-match-pattern match-pattern="^src/.*\.java" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src.*.*src-gen/.*\.java" include-pattern="false"/>
+    <file-match-pattern match-pattern="^src/main/resources.*\.properties" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src/test/resources.*\.properties" include-pattern="true"/>
+  </fileset>
 </fileset-config>
```
Comment (planger) [inline server/plugins/org.eclipse.glsp.ide.editor/.checkstyle:24]: I don't think this change is necessary.

File: server/plugins/org.eclipse.glsp.ide.editor/.classpath [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
-	<classpathentry kind="src" path="src/"/>
-	<classpathentry kind="output" path="target/classes"/>
-</classpath>
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/"/>
+	<classpathentry kind="output" path="target/classes"/>
+</classpath>
```
Comment (planger) [inline server/plugins/org.eclipse.glsp.ide.editor/.classpath:7]: I don't think this change is necessary.

File: server/example/org.eclipse.glsp.ide.workflow.editor/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -19,12 +20,16 @@ Require-Bundle: org.eclipse.elk.alg.common;bundle-version="0.8.1",
...
  org.eclipse.lsp4j.websocket;bundle-version="0.11.0",
  org.eclipse.e4.core.contexts;bundle-version="1.8.400",
- org.eclipse.jetty.server;bundle-version="[10.0.0,11.0.0)"
-Bundle-ClassPath: .
-Eclipse-BundleShape: dir
-Bundle-ActivationPolicy: lazy
-Bundle-Activator: org.eclipse.glsp.ide.workflow.editor.Activator
-Bundle-Vendor: Eclipse GLSP
+ org.eclipse.jetty.server;bundle-version="[10.0.0,11.0.0)",
+ org.eclipse.e4.ui.model.workbench;bundle-version="2.2.300.v20221024-2137",
```
Comment (planger) [inline server/example/org.eclipse.glsp.ide.workflow.editor/META-INF/MANIFEST.MF:24]: Do we need to put a specific bundle version here? At least it should be a range imho.

File: server/plugins/org.eclipse.glsp.ide.editor/src/org/eclipse/glsp/ide/editor/utils/UIUtil.java [hunk 6]
```diff
@@ -63,13 +62,7 @@ public static void asyncExec(final Runnable runnable) {
...
    }
 
-   public static Optional<Shell> findShell(final String clientId) {
-      return GLSPIdeEditorPlugin.getDefault().getGLSPEditorRegistry().getGLSPEditor(clientId)
-         .map(editor -> editor.getEditorSite().getShell())
-         .or(UIUtil::findShell);
-   }
-
-   private static Optional<Shell> findShell() {
+   public static Optional<Shell> findShell() {
```
Comment (planger) [inline server/plugins/org.eclipse.glsp.ide.editor/src/org/eclipse/glsp/ide/editor/utils/UIUtil.java:65]: Is there a reason why we remove `findShell(clientId)`? Wouldn't it be just as good to keep it and pass the `clientId` as before?

File: server/releng/org.eclipse.glsp.ide.feature/.checkstyle [hunk 7]
```diff
@@ -1,4 +1,14 @@
...
+    <property name="checkstyle.header.file" value="D:\dev\eclipse\eclipse-glsp-eclipse-integration\ws\.metadata\.plugins\org.eclipse.core.resources\.projects\org.eclipse.glsp.ide.feature\com.basistech.m2e.code.quality.checkstyleConfigurator\checkstyle-header-default.txt"/>
+    <property name="checkstyle.cache.file" value="${project_loc}/target/checkstyle-cachefile"/>
+  </local-check-config>
+  <fileset name="java-sources-default" enabled="true" check-config-name="maven-checkstyle-plugin default" local="true">
+    <file-match-pattern match-pattern="^src.*\.java" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src.*.*src-gen/.*\.java" include-pattern="false"/>
+    <file-match-pattern match-pattern="^src/main/resources.*\.properties" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src/test/resources.*\.properties" include-pattern="true"/>
+  </fileset>
```
Comment (planger) [inline server/releng/org.eclipse.glsp.ide.feature/.checkstyle:13]: I think this change isn't needed.

File: server/releng/org.eclipse.glsp.ide.repository/.checkstyle [hunk 8]
```diff
@@ -1,4 +1,14 @@
...
+    <property name="checkstyle.header.file" value="D:\dev\eclipse\eclipse-glsp-eclipse-integration\ws\.metadata\.plugins\org.eclipse.core.resources\.projects\org.eclipse.glsp.ide.repository\com.basistech.m2e.code.quality.checkstyleConfigurator\checkstyle-header-default.txt"/>
+    <property name="checkstyle.cache.file" value="${project_loc}/target/checkstyle-cachefile"/>
+  </local-check-config>
+  <fileset name="java-sources-default" enabled="true" check-config-name="maven-checkstyle-plugin default" local="true">
+    <file-match-pattern match-pattern="^src.*\.java" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src.*.*src-gen/.*\.java" include-pattern="false"/>
+    <file-match-pattern match-pattern="^src/main/resources.*\.properties" include-pattern="true"/>
+    <file-match-pattern match-pattern="^src/test/resources.*\.properties" include-pattern="true"/>
+  </fileset>
```
Comment (planger) [inline server/releng/org.eclipse.glsp.ide.repository/.checkstyle:13]: I think this change isn't needed.

**Review by planger — CHANGES_REQUESTED** [review]
Thanks again for this nice work! Most of what I tested seems to function nicely.

- For me on Ubuntu 22.04 the context menu didn't work though. I got the following exception:
```
java.lang.NullPointerException
	at org.eclipse.glsp.ide.editor.ui.BrowserContextMenuInstaller.requestContextMenu(BrowserContextMenuInstaller.java:45)
	at org.eclipse.glsp.ide.editor.ui.BrowserContextMenuInstaller$1.lambda$0(BrowserContextMenuInstaller.java:33)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:5040)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4520)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
```
- While undo/redo works, it isn't hooked up with the global undo/redo actions.
- The *Diagram* menu is not enabled with the e4 viewer

I also have some comments inline above and a few code-style suggestions which I added as a commit 816f80d82de63d92c4f775f71b1ecffa64f178dc.

File: server/example/org.eclipse.glsp.ide.workflow.editor/.checkstyle (hunk 1 above)
Comment (N1k145) [reply server/example/org.eclipse.glsp.ide.workflow.editor/.checkstyle:23]: Thanks, I don't know how this happened, but I noticed the same thing when I looked into the template repository you linked earlier. I revert those and the other changes like this

File: server/example/org.eclipse.glsp.ide.workflow.editor/META-INF/MANIFEST.MF (hunk 5 above)
Comment (N1k145) [reply server/example/org.eclipse.glsp.ide.workflow.editor/META-INF/MANIFEST.MF:24]: Missed this one, that does not really need any version

File: server/plugins/org.eclipse.glsp.ide.editor/src/org/eclipse/glsp/ide/editor/utils/UIUtil.java (hunk 6 above)
Comment (N1k145) [reply server/plugins/org.eclipse.glsp.ide.editor/src/org/eclipse/glsp/ide/editor/utils/UIUtil.java:65]: I removed it as I did not really see the reason for it. And there is no EditorSite anymore. I added it back by getting the Shell from the Browser Composite

**Comment by N1k145** [comment]
Thank you for the review.
I fixed your comments and included your changes.

With the undo redo action there is an issue in the Eclipse implementation, the Buttons are not retargetable to E4 Actions because of the implementation in org.eclipse.ui.internal.ide.WorkbenchActionBuilder.createEditMenu()
I will open an Issue and PR in platform.ui as we also had to build a workaround for this already.

**Comment by planger** [comment]
@N1k145 Just wanted to get back to you that we have your PR on our radar. We'll get back to it in the next couple of days. Thanks for your fixes regarding my earlier review!

**Comment by N1k145** [comment]
@planger thanks for the info, I'm on vacation next week and can't look into it during that time. So if something needs changes, I can have a look this week. Or after I'm back

**Review by planger — APPROVED** [review]
Thank you very much again for this contribution! Looks good to me!
I'll just let @tortmayr chime in if he has some further comments. Other than that, this is good to go. Thanks again!

**Review by tortmayr — APPROVED** [review]
Thank you very much for this contribution!
Nothing to add from my side. Code looks good to me and both the e3 and e4 implementations work as expected. 👍🏼

**Comment by N1k145** [comment]
@planger @tortmayr thank you for the reviews and merging this PR

**Pull request merged**